### ISSUE TITLE
Forge Port

### ISSUE THREAD

**Issue by qixils** [qixils/minecraft-crowdcontrol#146]
# TL;DR

- The mod currently recommends the Fabric mod loader. It supports Fabric on 1.19.2 and above. The Fabric ecosystem is rich with new mods and developer tools being created for and ported to it everyday.
- The mod also supports Forge via SpongeForge. SpongeForge supports 1.12.2 and 1.16.5 with plans in the [near-ish future](https://github.com/orgs/SpongePowered/projects/26/views/3) to support 1.19.4 and 1.20.5. (**EDIT**: [SpongeForge 1.19.4 is officially in beta](https://spongepowered.org/downloads/spongeforge?minecraft=1.19.4&offset=0)! Work is under way to bring support to this version.)
- To create a native Forge port independent of Sponge would require [several libraries](#libraries) to release Forge ports, but the chances of this are unlikely due to various complicated factors and the time/effort to do this work alone would be considerable.
- A native Forge port will not be created for old versions of the game.

# The Full Situation

The mod supports the Sponge API, and Sponge can run as a Forge mod, so the mod supports Forge, right? Well, yes, but only on the Minecraft versions supported by SpongeForge, which currently is just 1.12.2 and 1.16.5 (edit: and 1.19.4!). This is great for speedrunners of old versions (which honestly was my main goal) but also for some old modpacks. So, what about Forge for the latest version of the game? Well.

## The Native Forge Port

Over a year ago when I began development of the Fabric mod, I planned to release a Forge version alongside it, but this quickly proved to be nonviable. I'd have to singlehandedly create Forge ports of several software libraries (listed below) that may or [likely] may not even want a Forge port created of them, meaning I'd likely have to maintain several of them in perpetuity. It's a lot to ask for just to support a modding platform that is increasingly disfavored by modders and players for its [toxic](https://github.com/CaffeineMC/caffeine-meta/wiki/FAQ#where-are-the-forge-versions-of-your-mods) [management](https://quiltmc.org/en/blog/2023-04-26-the-statement/#forge), slow updates, lack of support for snapshots and [April 1st updates](https://github.com/qixils/minecraft-crowdcontrol/tree/fabric/23w13a_or_b#readme), disinterest in supporting game code patches (which this mod makes good use of), et cetera. I don't honestly anticipate a native Forge port coming to fruition for these reasons, but regardless, the theoretical path to native Forge support is laid out below.

EDIT: The newly released NeoForge appears to be tackling several of these issues and attracting new mod ports, but the majority of the libraries we use seem to remain uninterested in supporting it.

### Libraries

First, the following libraries used by the Fabric mod would need to be ported to Forge.

- [ ] [adventure-platform](https://github.com/KyoriPowered/adventure-platform-fabric)
- [ ] [cloud](https://github.com/Incendo/cloud)
- [ ] [Cardinal Components API](https://github.com/OnyxStudios/Cardinal-Components-API)
	- [ ] Relevant: #147
- [x] [Cloth Config API](https://github.com/shedaniel/cloth-config)

To support older versions of the game, these theoretical Forge ports would also need to be backported to old versions of the game which is even less likely to happen.

### Porting the Fabric Mod to Forge

Second, since the Forge and Fabric mods would mostly share the same code, the code base for the Fabric mod would have to be modularized to allow the development of the Forge mod. Finally, the Forge mod initializer would have to be created.

- [ ] Split module back out to a common mojmap module, a fabric initializer module, and a forge initializer module
  - [Architectury](https://github.com/architectury) may be of help for this, namely for its easy mixin support
- [ ] Expose GUI config on Forge
- [ ] ???
- [ ] Test Fabric and Forge extensively

## The Sponge Forge Port

There's a high chance that the native Forge port never comes to fruition. In this case, the next best hope would be for Sponge to finally bring Forge support to the latest version of the game, and fortunately they [seem to be looking into doing just that in the near-ish future](https://github.com/orgs/SpongePowered/projects/26/views/3). If SpongeForge ever releases a build for the latest version of the game and this mod runs on it, I'll likely consider this issue resolved.

## The Future

I'm always looking around the Minecraft development space for new platforms to support and that includes Forge. I'll be posting comments with updates as I have them so do feel free to subscribe to this issue if you'd like to hear about future developments.

**Comment by qixils** [comment]
Much ado has recently been made about NeoForge, a fork of Forge created by most of the former Forge team that excludes one of the more toxic team members and supposedly sheds some of the project's negative feelings towards mods utilizing game code patches. I believe this project may make more mods and libraries willing to adopt Forge, though I think it remains unlikely that many people will put in the effort to do so.

**Comment by qixils** [comment]
Sponge has finally published [stable builds of SpongeForge 1.16.5](https://spongepowered.org/downloads/spongeforge?minecraft=1.16.5&offset=0) to mark the end of active SF development on 1.16.5 and to hopefully mark the beginning of active SF development on 1.19.4+.

**Comment by qixils** [comment]
Recent-ish comment from a Sponge developer regarding the status of SpongeForge:
> I suspect the next version will be 1.20.2 for SF
> we're sorta hoping to let the dust settle on the whole forge/neoforge situation too

**Comment by qixils** [comment]
Even better: on the [most recent State of Sponge](https://youtu.be/IvW2ieE3bg8?t=401), the devs announced that SpongeForge 1.19.4 is [in unstable alpha](https://github.com/SpongePowered/Sponge/actions/workflows/build-project.yaml?query=is%3Asuccess+event%3Apush+branch%3Aapi10%2Fforge) with plans to officially publish beta releases very soon.

**Comment by qixils** [comment]
Just recently, Sponge began shipping beta builds of SpongeForge for the latest version of the game, 1.20.6. The mod has successfully been updated to fully support this version with minor changes, and so, as promised in the original issue, this issue will now be closed.

**Comment by qixils** [comment]
adventure-platform-fabric, the biggest blocker in porting to Forge, has started working on support for NeoForge. The v3.6.1 release of CC will migrate away from the Cardinal Components library, leaving adventure-platform-fabric as the only library blocking support for NeoForge. We will continue to monitor the situation.

**Comment by qixils** [comment]
🤔

![image](https://github.com/user-attachments/assets/83e56833-2623-4292-899c-817599d6342d)
